File: Universe_Backend/Repositories/IUsersRepository.cs
using Universe_Backend.Data.DTOs;
using Universe_Backend.Data.Models;

namespace Universe_Backend.Repositories;

public interface IUsersRepository
{
    public Task<IEnumerable<UserDTO>> SearchUsers(string query, DateTime? lastDate, string? lastId);
    public Task<UserDTO> GetUser(string id);
    public Task AddFollower(string followerId, string followedId);
    public Task RemoveFollower(string followerId, string followedId);
    public Task<IEnumerable<UserDTO>> GetFollowers(string userId, DateTime? lastDate, string? lastId);
    public Task<IEnumerable<UserDTO>> GetFollowing(string userId, DateTime? lastDate, string? lastId);
    public Task UpdateUser(UserDTO user);
    public Task<int> GetFollowersCount(string userId);
    public Task<int> GetFollowingCount(string userId);
    public Task<IEnumerable<NotificationToken>> GetNotificationTokens(string userId);
    public Task<bool> IsUserNameAvailable(string userName);
    public Task<bool> IsUserOneFollowingUserTwo(string userOneId, string userTwoId);
}

File: universe/lib/interfaces/iusers_repository.dart
import 'package:universe/models/data/user.dart';
import 'package:universe/models/requests/users_api_call_start.dart';

abstract class IUsersRepository {
  Future<User> getUser(String userId);

  Future<List<User>> searchUsers(
      String query, UsersApiCallStart start, int limit);

  Future<List<User>> getFollowers(String userId, UsersApiCallStart start);

  Future<List<User>> getFollowing(String userId, UsersApiCallStart start);

  Future followUser(String followedId, String followerId);

  Future unfollowUser(String followedId, String followerId);

  Future<int> getFollowersCount(String userId);

  Future<int> getFollowingCount(String userId);

  Future updateUser(String userId, User user);

  Future<bool> isUsernameAvailable(String username);

  Future<bool> isFollowing(String followerId, String followedId);
}
